File: Maximum Profit By Choosing A Subset Of Intervals.cpp
class Solution {
  public:
    int find_nearest_equal_or_greater(int start, int end, int value, int n,
                                      vector<vector<int>> &intervals) {

        // If nothing found, position will be n.
        int position = n;
        while (start <= end) {
            int mid = (start + end) / 2;
            if (intervals[mid][0] >= value) {
                position = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        return position;
    }

    int maximum_profit_helper(int index, int n, vector<vector<int>> &intervals,
                              vector<int> &dp) {
        if (index >= n) {
            return 0;
        }

        // `dp[index]` stores the maximum profit one can get by choosing
        // non-overlapping intervals from `index` to `n - 1`.
        if (dp[index] != -1) {
            return dp[index];
        }

        int answer = maximum_profit_helper(index + 1, n, intervals, dp);
        int new_index = find_nearest_equal_or_greater(
            index + 1, n - 1, intervals[index][1], n, intervals);
        answer =
            max(answer, intervals[index][2] +
                            maximum_profit_helper(new_index, n, intervals, dp));

        return dp[index] = answer;
    }

    int maximum_profit(int n, vector<vector<int>> &intervals) {
        sort(intervals.begin(), intervals.end());
        vector<int> dp(n, -1);
        return maximum_profit_helper(0, n, intervals, dp);
    }
};

File: Maximum Profit By Choosing A Subset Of Intervals.py
from typing import List

maxi = 10**5+2
class Solution:
    def maximum_profit(self, n : int, intervals : List[List[int]]) -> int:
        intervals.sort(key = lambda x:x[1])
        
        dp = [0]*maxi
        
        ind = 0
        i = 1
        while i < maxi:
            dp[i] = max(dp[i], dp[i-1])
            if ind < n and intervals[ind][1] == i:
                st, en, val = intervals[ind]
                dp[i] = max(dp[i], dp[st]+val)
                ind += 1
                continue
            i += 1
        return dp[maxi-1]
